libcudacxx/include/cuda/__floating_point/fp.h

```diff
@@ -0,0 +1,530 @@
+#  pragma system_header
+#endif // no system header
+
+#if _CCCL_STD_VER >= 2017
+
+#  include <cuda/__floating_point/config.h>
+#  include <cuda/__floating_point/conv_rank_order.h>
+#  include <cuda/__floating_point/storage.h>
+#  include <cuda/__floating_point/type_traits.h>
+#  include <cuda/__fwd/fp.h>
+#  include <cuda/std/__bit/bit_cast.h>
+#  include <cuda/std/__bit/integral.h>
+#  include <cuda/std/__concepts/concept_macros.h>
+#  include <cuda/std/__cstddef/types.h>
+#  include <cuda/std/__type_traits/enable_if.h>
+#  include <cuda/std/__type_traits/is_constant_evaluated.h>
+#  include <cuda/std/__type_traits/is_integral.h>
+#  include <cuda/std/__type_traits/is_same.h>
+#  include <cuda/std/__type_traits/make_signed.h>
+#  include <cuda/std/__type_traits/remove_cv.h>
+#  include <cuda/std/climits>
+
+#  include <nv/target>
+
+// Silence the warning about the use of long double in device code
+_CCCL_NV_DIAG_SUPPRESS(20208)
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+// Signed FP types
+using fp4_e2m1 = __fp<__fp4_e2m1_config>;
+using fp6_e2m3 = __fp<__fp6_e2m3_config>;
+using fp6_e3m2 = __fp<__fp6_e3m2_config>;
+using fp8_e4m3 = __fp<__fp8_e4m3_config>;
+using fp8_e5m2 = __fp<__fp8_e5m2_config>;
+using fp16     = __fp<__fp16_config>;
+using bf16     = __fp<__bf16_config>;
+using fp32     = __fp<__fp32_config>;
+using fp64     = __fp<__fp64_config>;
+// #  if !defined(_LIBCUDACXX_HAS_NO_INT128)
+// using fp128 = __fp<__fp128_config>;
+// #  endif // !_LIBCUDACXX_HAS_NO_INT128
+
+// Unsigned FP types
+using fp8_ue4m3 = __fp<__fp8_ue4m3_config>;
+using fp8_ue8m0 = __fp<__fp8_ue8m0_config>;
+
+_CCCL_TEMPLATE(class _Tp)
+_CCCL_REQUIRES(__fp_is_floating_point_v<_Tp>)
+__fp(_Tp) -> __fp<__fp_make_config_from_t<_Tp>>;
+
+template <class _Tp>
+__fp(__fp_from_native_t, _Tp) -> __fp<__fp_make_config_from_t<_Tp>>;
+
+template <class _To, class _From>
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr bool __fp_cast_is_implicit()
+{
+  constexpr auto __rank_order = ::cuda::__fp_make_conv_rank_order<_To, _From>();
+
+  return __rank_order == __fp_conv_rank_order::__equal || __rank_order == __fp_conv_rank_order::__greater;
+}
+
+template <class _Config>
+class __fp
+{
+public:
+  template <class>
+  friend class __fp;
+  friend _Config;
+  friend struct __fp_cast;
+  friend struct __fp_ops;
+
+  using __config_type = _Config;
+
+  static constexpr size_t __exp_nbits  = __config_type::__exp_nbits;
+  static constexpr size_t __mant_nbits = __config_type::__mant_nbits;
+  static constexpr bool __is_signed    = __config_type::__is_signed;
+  static constexpr size_t __nbits      = __exp_nbits + __mant_nbits + __is_signed;
+  static constexpr bool __has_inf      = __config_type::__has_inf;
+  static constexpr bool __has_nan      = __config_type::__has_nan;
+  static constexpr bool __has_nans     = __config_type::__has_nans;
+  static constexpr bool __has_denorm   = __config_type::__has_denorm;
+
+  using __storage_type       = __fp_storage_t<__nbits>;
```

Comment:
This could be `cuda::std::bitset` - it means there's some extra logic for extracting the bits at times, but it would allow for _some_ operations (that don't need to extract mantissas, really) to still work on floating point types larger than the largest integer type available. (Once again, you can see how I dealt with some of the operations on my branch, in representation.h.)

Reply:
This would need to be a specialization for types larger than the largest integer type.

The reason being that we really want to keep compile times down and bitset is not helping with that